File: in_place.ml
let swap arr i j =
  let temp = arr.(i) in
  arr.(i) <- arr.(j);
  arr.(j) <- temp

let partition arr lo hi =
  let pivot = arr.(hi) in
  let i = ref lo in
  for j = lo to hi - 1 do
    if arr.(j) < pivot then (
      swap arr !i j;
      incr i
    )
  done;
  swap arr !i hi;
  !i

let rec qsort arr lo hi =
  if lo < hi then (
    let p = partition arr lo hi in
    qsort arr lo (p - 1);
    qsort arr (p + 1) hi
  )

let quicksort arr =
  if Array.length arr > 0 then
    qsort arr 0 (Array.length arr - 1)

let () =
  let verbose = Array.mem "-v" Sys.argv in
  let pre_print = Array.mem "-p" Sys.argv in
  
  (* Generate test data *)
  let xs = Array.init 10000 (fun i -> (i * 73 + 19) mod 10000) in
  
  if pre_print then (
    print_endline "Before sorting:";
    print_string "[";
    Array.iteri (fun i x ->
      if i > 0 then print_string ", ";
      print_int x
    ) xs;
    print_endline "]"
  ) else (
    quicksort xs;
    if verbose then (
      print_string "[";
      Array.iteri (fun i x ->
        if i > 0 then print_string ", ";
        print_int x
      ) xs;
      print_endline "]"
    ) else (
      print_string "[";
      for i = 0 to 9 do
        if i > 0 then print_string ", ";
        print_int xs.(i)
      done;
      print_endline "]"
    )
  )

File: in_place.cpp
#include <iostream>
#include <vector>
#include <string>

void qsort(std::vector<int>& a, int lo, int hi);

int partition(std::vector<int>& a, int lo, int hi) {
    int pivot = a[hi];
    int i = lo;
    for (int j = lo; j < hi; ++j) {
        if (a[j] < pivot) {
            std::swap(a[i], a[j]);
            ++i;
        }
    }
    std::swap(a[i], a[hi]);
    return i;
}

void qsort(std::vector<int>& a, int lo, int hi) {
    if (lo >= hi) return;
    int p = partition(a, lo, hi);
    qsort(a, lo, p - 1);
    qsort(a, p + 1, hi);
}

int main(int argc, char* argv[]) {
    bool verbose = false;
    bool prePrint = false;
    for (int i = 1; i < argc; ++i) {
        if (std::string(argv[i]) == "-v") {
            verbose = true;
        } else if (std::string(argv[i]) == "-p") {
            prePrint = true;
        }
    }
    std::vector<int> xs(10000);
    for (int i = 0; i < 10000; ++i) {
        xs[i] = (i * 73 + 19) % 10000;
    }
    if (prePrint) {
        std::cout << "Before sorting:\n";
        for (int v : xs) {
            std::cout << v << " ";
        }
        std::cout << "\n";
        return 0;
    }
    if (!xs.empty()) {
        qsort(xs, 0, (int)xs.size() - 1);
    }
    if (verbose) {
        for (int v : xs) {
            std::cout << v << " ";
        }
        std::cout << "\n";
    } else {
        for (int i = 0; i < 10; ++i) {
            std::cout << xs[i] << " ";
        }
        std::cout << "\n";
    }
}
